docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,627 @@
+
+    ones = torch.ones([2])
+    m = torch.jit.script(f)
+    print(m(ones))
+
+The above example produces the following output
+
+.. testoutput::
+
+    1
+    1
+    [ CPUFloatType{2} ]
+    True
+
+Primitive Types
+^^^^^^^^^^^^^^^
+
+Primitive TorchScript types represent types that represent a single type of value and go with a single pre-defined
+type name.
+
+::
+
+    TSPrimitiveType := "int" | "float" | "double" | "complex" | "bool" | "str" | "None"
+
+Structural Types
+^^^^^^^^^^^^^^^^
+
+Structural types are types that are structurally defined without a user-defined name (unlike nominal types),
+such as ``Future[int]``. Structural types are composable with any ``TSType``.
+
+::
+
+    TSStructualType :=  TSTuple | TSNamedTuple | TSList | TSDict |
+                        TSOptional | TSFuture | TSRRef
+
+    TSTuple := "Tuple" "[" (TSType ",")* TSType "]"
+    TSNamedTuple := "namedtuple" "(" (TSType ",")* TSType ")"
+    TSList := "List" "[" TSType "]"
+    TSOptional := "Optional" "[" TSType "]"
+    TSFuture := "Future" "[" TSType "]"
+    TSRRef := "RRef" "[" TSType "]"
+    TSDict := "Dict" "[" KeyType "," TSType "]"
+    KeyType := "str" | "int" | "float" | "bool" | TensorType | "Any"
+
+where
+
+* Tuple, List, Optional, Union, Future, Dict represent Python type class names defined in module typing. Therefore before using these type names, one must import them from typing (e.g., from typing import Tuple).
+* namedtuple represents Python class  collections.namedtuple or typing.NamedTuple .
+* Future and RRef represent Python classes  torch.futures, torch.distributed.rpc.
+
+Compared to Python
+""""""""""""""""""
+
+Apart from only composable with TorchScript types, these TorchScript structural types often support a common subset of
+the operators and methods of its Python counterparts.
+
+Example:
+
+This example uses typing.NamedTuple syntax:
+
+.. testcode::
+
+    import torch
+    from typing import NamedTuple
+    from typing import Tuple
+
+    class MyTuple(NamedTuple):
+        first: int
+        second: int
+
+    def inc(x: MyTuple) -> Tuple[int, int]:
+        return (x.first+1, x.second+1)
+
+    t = MyTuple(first=1, second=2)
+    scripted_inc = torch.jit.script(inc)
+    print("TorchScript:", scripted_inc(t))
+
+This example uses ``collections.namedtuple`` syntax:
+
+.. testcode::
+
+    import torch
+    from typing import NamedTuple
+    from typing import Tuple
+    from collections import namedtuple
+
+    _AnnotatedNamedTuple = NamedTuple('_NamedTupleAnnotated', [('first', int), ('second', int)])
+    _UnannotatedNamedTuple = namedtuple('_NamedTupleAnnotated', ['first', 'second'])
+
+    def inc(x: _AnnotatedNamedTuple) -> Tuple[int, int]:
+        return (x.first+1, x.second+1)
+
+    m = torch.jit.script(inc)
+    print(inc(_UnannotatedNamedTuple(1,2)))
+
+Example:
+
+This example illustrates a common mistake of annotating structural types, i.e., not importing the composite type
```

Comment:
IMHO, this example doesn't quite fit here, this is part of spec of TorchScript. Rather, Python 3 dictates Tuple annotation is imported and used this way.

Reply:
I am not sure where else this information should be added. I can probably add saying that it's Python 3  requirement?

Reply:
Just thought more about it, we can keep it here, I am fairly sure someone will find it useful.